core: fix test failures in test/core/validation.js

Failing file
- `test/core/validation.js`

Example failure
- FAIL: Core htmx client side validation tests > Validation skipped for indirect form submission
- AssertionError: expected 'No Request' to equal 'Clicked!'

Repro
1. Ensure no `!test/...` excludes are active in `web-test-runner.config.mjs`.
2. Run: `pnpm exec wtr --config web-test-runner.config.mjs --files test/core/validation.js --concurrency 1 --static-logging`

Goal
- Make this file pass in `pnpm test:htmx`.

--- a/cmd/htmx_test/main.mbt
+++ b/cmd/htmx_test/main.mbt
@@ -64,9 +64,23 @@ extern "js" fn dispatch_event(
 ) -> Unit =
   #|(target, event_name, detail) => {
   #|  if (!target) return;
-  #|  const opts = { bubbles: true, cancelable: true };
-  #|  if (detail != null) opts.detail = detail;
-  #|  const evt = detail != null ? new CustomEvent(event_name, opts) : new Event(event_name, opts);
+  #|  // Always use CustomEvent with detail to ensure compatibility with hyperscript
+  #|  // hyperscript expects event.detail to exist, even if empty
+  #|  // If detail is null or undefined, create an empty object
+  #|  // If detail exists but doesn't have 'elt', add target as 'elt' for htmx events
+  #|  let finalDetail = detail != null ? detail : {};
+  #|  // Ensure htmx-related events have 'elt' property pointing to target
+  #|  if (event_name.startsWith('htmx:') || eventName.startsWith('htmx:')) {
+  #|    if (!finalDetail.elt) {
+  #|      finalDetail = { ...finalDetail, elt: target };
+  #|    }
+  #|  }
+  #|  const opts = {
+  #|    bubbles: true,
+  #|    cancelable: true,
+  #|    detail: finalDetail
+  #|  };
+  #|  const evt = new CustomEvent(event_name, opts);
   #|  target.dispatchEvent(evt);
   #|}
 

--- a/src/htmx/processor.mbt
+++ b/src/htmx/processor.mbt
@@ -732,18 +732,29 @@ fn handle_click(evt : @core.Any) -> Unit {
   if is_submit_button(target_el) {
     if has_containing_form(target_el) {
       let form = find_closest_form(target_el)
-      // Check if form itself has an htmx method attribute
-      match find_method_url(form) {
+      // Check if the button itself has an htmx method attribute
+      // If so, it should handle the click, not the form (indirect form submission)
+      match find_method_url(target_el) {
         Some(_) => {
-          // Submit buttons should always submit the form, regardless of hx-trigger
-          // The form's hx-trigger only affects what events automatically trigger,
-          // not explicit user action of clicking a submit button
-          event.preventDefault()
-          // Pass the button as submitter for formnovalidate check
-          process_element_with_trigger(form, Some(target_el))
-          return
+          // Button has its own htmx request, skip form submit handling
+          // It will be processed by find_htmx_element below
+          ()
         }
-        None => ()
+        None => {
+          // Button doesn't have htmx, check if form does
+          match find_method_url(form) {
+            Some(_) => {
+              // Submit buttons should always submit the form, regardless of hx-trigger
+              // The form's hx-trigger only affects what events automatically trigger,
+              // not explicit user action of clicking a submit button
+              event.preventDefault()
+              // Pass the button as submitter for formnovalidate check
+              process_element_with_trigger(form, Some(target_el))
+              return
+            }
+            None => ()
+          }
+        }
       }
     }
   }

--- a/src/htmx/validation.mbt
+++ b/src/htmx/validation.mbt
@@ -29,7 +29,11 @@ extern "js" fn dispatch_validate_events(form : @dom.Element) -> Unit =
   #|(form) => {
   #|  const inputs = form.querySelectorAll('input, select, textarea');
   #|  for (let i = 0; i < inputs.length; i++) {
-  #|    const evt = new CustomEvent('htmx:validation:validate', { bubbles: true, cancelable: true });
+  #|    const detail = {
+  #|      elt: inputs[i],
+  #|      validity: inputs[i].validity || {}
+  #|    };
+  #|    const evt = new CustomEvent('htmx:validation:validate', { bubbles: true, cancelable: true, detail: detail });
   #|    inputs[i].dispatchEvent(evt);
   #|  }
   #|}
@@ -38,7 +42,11 @@ extern "js" fn dispatch_validate_events(form : @dom.Element) -> Unit =
 /// Dispatch htmx:validation:validate event on a single element
 extern "js" fn dispatch_validate_event(el : @dom.Element) -> Unit =
   #|(el) => {
-  #|  const evt = new CustomEvent('htmx:validation:validate', { bubbles: true, cancelable: true });
+  #|  const detail = {
+  #|    elt: el,
+  #|    validity: el.validity || {}
+  #|  };
+  #|  const evt = new CustomEvent('htmx:validation:validate', { bubbles: true, cancelable: true, detail: detail });
   #|  el.dispatchEvent(evt);
   #|}
 

